### PR TITLE
IPv6: Ensure calculated node CIDR size for pod subnets is valid

### LINKED ISSUE #58766: IPv6: Pod net CIDR size calculated is invalid
**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**What happened**:
For IPv6, the node CIDR size is hard coded to /64. This causes an error bringing up the kube-controller, as the size must be for a smaller subnet (larger value) that what is used for the pod subnet, which is specified for some CNI plugins (e.g. Calico).

In addition, this is a catch-22, as the pod subnet prefix cannot be for a larger network (smaller value) than /66, so we cannot set it to /63, for example.

**What you expected to happen**:
The node CIDR size should be computed, based on the pod subnet specified, so that it can meet the requirements. 

**How to reproduce it (as minimally and precisely as possible)**:
Bring up a cluster using KubeAdm, specifying IPv6 and choose any pod subnet prefix desired.

**Anything else we need to know?**:
Note, there is a second requirement, that the difference between pod subnet size and node CIDR size be 16 or less (issue 44918).

I'll be submitting a proposed PR to address this.

**Environment**:
- Kubernetes version (use `kubectl version`): latest on master, using 1.9.2 in my case
- Cloud provider or hardware configuration: bare metal, single node
- OS (e.g. from /etc/os-release): Ubuntu 16.04
- Kernel (e.g. `uname -a`):
Linux bxb-c2-77 4.4.0-109-generic #132-Ubuntu SMP Tue Jan 9 19:52:39 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

- Install tools: kubeadm
- Others:
/area ipv6
/sig network

### PR DESCRIPTION
**Pull request by pmichali**
With IPv4, the node CIDR prefix is set to /24, which gives 256 pods per node
and 256 nodes, when assuming a /16 is used for the pod subnet.

For IPv6, the node CIDR prefix, is hard coded to /64. This does not work,
because the pod subnet prefix must be /66 or higher and must be a larger subnet
(lower value) than the node CIDR prefix.

In addition, the bit mask used to track the subnets (implying the number of
nodes), can only handle 32K entries, so the difference between pod subnet
prefix and node CIDR prefix cannot be more than 16 (bits).

To address this, the following algorithm is proposed to provide as many pods
per node as possible, and not exceed the number of nodes.

If the pod subnet prefix is from /66 to /104, we'll set the node CIDR prefix
to 16 more, so that the bit map is not exceeded. The rest of the bits will be
for pods per node.

If the subnet prefix is from /105 to /112, we'll split the available bits
between what is used for the nodes and what is used for the pods per node.
This will give a node CIDR prefix from /116 to /120.

If the subnet prefix is from /113 to /119, we'll do like IPv4 and ensure that
there are 256 pods per node, and the remaining bits will be for the nodes,
giving a node CIDR prefix of /120 always. This supports a limited number of
nodes, in some cases.

If the subnet prefix is /120 to /128, we don't have enough bits and will set
the node CIDR prefix to be the same as the pod subnet prefix. This will cause
a falure later, when it tests that the pod subnet prefix is larger than the
node CIDR prefi.

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #58766 

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by pmichali** [comment]
Fixes #58766

**Comment by pmichali** [comment]
Whoops... used some Python variable naming... fixing.

**Comment by timothysc** [comment]
@errordeveloper or @kad - would you have time to review this one?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 1]
```diff
@@ -228,6 +229,48 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
 	return command
 }
 
+// calcNodeCidrSize determines the size of the subnets used on each node, based
+// on the pod subnet provided.  For IPv4, we assume that the pod subnet will
+// be /16 and use /24. For IPv6, there are some restrictions. First, the pod
+// subnet size must be /66 or larger. Second, the bit mask used for allocating
+// subnets (one per node) allows up to 32K (16 bits). To accomodate this, the
+// following algorithm is used, based on pod subnet size (implying the number of
+// bits available), to determine the node CIDR size.
```
Comment (timothysc) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Doesn't this need to all feed into the pod-per-node configuration?

Comment (pmichali) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Can you elaborate? Is there some additional work needed? If so, what (as I'm not that familiar with this code)?

From what I can tell, in the existing code, for IPv4, it hard codes the node CIDR to /24 giving 256 pods per node.

This logic that I've introduced, leaves IPv4 unchanged (/24), and attempts to do the same for IPv6, but accommodating for additional constraints.

It'll assign 256 pods per node, when small subnets are used, but as the subnets get larger, it'll split the allocation of pods per node and nodes for the cluster to give > 256 pods per node. When the subnets are very large, we must limit the number of nodes supported (32K), and give the rest for pods per node.

It's expected that very large subnets will be used in practice, with well more than 256 pods per node allowed (and we must, to not violate the constraint for max number of nodes).

As it stands right now, it is totally broken. No subnet size will work.

**Comment by errordeveloper** [comment]
I don't have enough expertise in IPv6 area wrt Kubernetes code. Maybe
@mikebryant could help?

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 2]
```diff
@@ -228,6 +229,48 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
...
+//
+// POD SUBNET SIZE   BITS AVAIL   ACTION
+//    /66  to /104   62 to 24     Allow 16 bits for node CIDR, rest for pods/node
+//    /105 to /112   16 to 23     Give node CIDR half of the bits available
+//    /113 to /119   15 to  9     Give 8 bits for pods/node and rest for node CIDR
+//    /120 to /128   --------     Set node CIDR to pod subnet side - will be rejected later
+//
```
Comment (kad) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Algorithm seems to be ok, but in general, I have doubts that somebody will be trying to fit whole datacenter in single /64 (which is considered minimally viable allocation for interface range in most of the routers).
From aesthetic appearance, if we are about to split available bits between node and pod, it will be good to round it to 16bit for pods. This way by just looking at IPv6 address it will be possible to say which node it belongs to. 
e.g. 

```
2001:0DB8:1234:1234:1234:1234:1234:abcd/112 = pod ip
2001:0DB8:1234:1234:1234:1234:1234::/112 = node range.
```

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go [hunk 3]
```diff
@@ -661,6 +661,72 @@ func TestGetControllerManagerCommand(t *testing.T) {
...
+			expectedPrefix: "24",
+		},
+		{
+			name:           "V4: Always uses 24",
+			podSubnet:      "10.10.10.10/16",
+			expectedPrefix: "24",
+		},
+		{
+			name:           "V6: Use pod subnet size, when not enough space",
+			podSubnet:      "fd00::2/128",
```
Comment (kad) [inline cmd/kubeadm/app/phases/controlplane/manifests_test.go]: better to use `2001:db8::/32` range for tests.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (timothysc) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: This change is fine, but we need to jig max pods/node with CIDR, which can be done separately.  I also think upping the default to something sane ~256 seems reasonable and all the defaults would just work.   @yujuhong

File: cmd/kubeadm/app/phases/controlplane/manifests.go [hunk 4]
```diff
@@ -228,6 +229,48 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
...
+// be /16 and use /24. For IPv6, there are some restrictions. First, the pod
+// subnet size must be /66 or larger. Second, the bit mask used for allocating
+// subnets (one per node) allows up to 32K (16 bits). To accomodate this, the
+// following algorithm is used, based on pod subnet size (implying the number of
+// bits available), to determine the node CIDR size.
+//
+// POD SUBNET SIZE   BITS AVAIL   ACTION
+//    /66  to /104   62 to 24     Allow 16 bits for node CIDR, rest for pods/node
+//    /105 to /112   16 to 23     Give node CIDR half of the bits available
+//    /113 to /119   15 to  9     Give 8 bits for pods/node and rest for node CIDR
```
Comment (pmichali) [inline cmd/kubeadm/app/phases/controlplane/manifests.go]: Note: With /119, we can only have one node, with 256 pods. Should that range be changed to /113 - /118, and consider /119 an invalid configuration, or reduce the number of pods/node to 128 for /119?

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 2 above)
Comment (pmichali) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Let me try to clarify the context here. This is not for the entire data center. It is for the pod subnets in the Kubernetes, where the cluster is set up with a **configured** pod network CIDR. From that CIDR, subnets will be created for each node.

For example, if the pod network CIDR was configured to be fd00::/112, then we have 16 bits available. With the above algorithm, 8 bits would be reserved for pods per node, and 8 bits for the number of nodes in the cluster.

The algorithm I chose, tries to balance between a minimum number of pods per node (256+), and keep within the limit for maximum network size (32k nodes). It uses three ranges. For small pod network configurations, it uses 256 pods/node. For large pod networks, it sets number of nodes to max (32k), and the remainder to pods/network. In between these two blocks, it just splits the number of available bits, so that we could have more than 256 pods/node.

Another scheme would be to try to maximize the number of pods/node, using multiples of 8 or 16, giving the rest to the number of nodes in the cluster. That does make the algorithm more complex, as we'd want to have at least 2 nodes in the cluster.  For a multiple of 8, algorithm, we'd have mappings...

/66 - /70 use /72 giving 56 bits for pods/node and 2-6 bits nodes
/71 - /78 use /80 giving 48 bits for pods/node and 2-9 bits for nodes
/79 - /86 use /88 giving 40 bits for pods/node and 2-9 bits for nodes
/87 - /94 use /96 giving 32 bits for pods/node and 2-9 bits for nodes
...

Is there desire to maximize the number of pods per node?

The key point in all this, is that the pod network CIDR is a configured value, and it would be to the benefit of the cluster to choose a "good" value, if readability is important.

File: cmd/kubeadm/app/phases/controlplane/manifests_test.go [hunk 5]
```diff
@@ -645,8 +645,8 @@ func TestGetControllerManagerCommand(t *testing.T) {
 				"--use-service-account-credentials=true",
 				"--controllers=*,bootstrapsigner,tokencleaner",
 				"--allocate-node-cidrs=true",
-				"--cluster-cidr=2001:101:115::/48",
-				"--node-cidr-mask-size=64",
+				"--cluster-cidr=2001:101:115::/80",
+				"--node-cidr-mask-size=96",
```
[the remark this answers is not in the document]
Comment (pmichali) [reply cmd/kubeadm/app/phases/controlplane/manifests_test.go]: That is not possible, as the pod network CIDR must be /66 or larger and the node CIDR must be larger than the pod network CIDR. Hence the issue with the existing code, it is forcing node CIDR to /64, but pod network must be /66 or larger and we are in a catch-32 (or maybe catch-64 for IPv6 :)

We can switch this IP to a documentation IP, if desired.

**Review by pmichali — COMMENTED** [review]
Thanks for the comments @kad and @timothysc. See responses in-line.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 4 above)
Comment (pmichali) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Correction... with /119 we could have two nodes, with 256 pods. Would that limitation be OK, or would we want to require /118 as the limit or restrict the number of pods/node to 128 in that case?

**Review by timothysc — APPROVED** [review]
The change is ok by me, but I'll defer to @kad for final review. 

/approve

**Comment by pmichali** [comment]
@kad Three questions from the comments?
1) Do we want to change the algorithm from a balanced approach to a pod/node priority?
2) If so, do we want to have node bits range from 2-9 or 1-8 (allow minimum of four or two nodes)?
3) For current scheme, do we want to allow /119 with two nodes and 256 pods/node - as currently setup, allow /119 with four nodes and 128 pods/node (special case), or only allow /118 with four nodes and 256 pods/node as the maximum allowed pod network CIDR?

Note: item #3 is a real edge case, as I can't image anyone using a pod network CIDR that large.

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (yujuhong) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: I didn't get the need of changing the (default) max pods per node. The cap should depend on the machine capacity, as well as the workloads.

Comment (pmichali) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: @yujuhong For IPv4, it is using 256 pods per node. 

For IPv6, I tried to do the same thing, however, when the subnets get larger, we are restricted by the number of nodes that can be supported (32K). So, if the subnet is between /66 and /103, we must allow a larger number of pods per node (128-103=25-16=9, 128-66=62-16=46).

Granted, this would be how many pods are **allowed** per node, but that doesn't mean we need to use them.

**Comment by kad** [comment]
Whole reason why I'm raising those questions are about generic practices on how IPv6 address allocations are supposed to be used. They are different than in IPv4. In IPv4 case, it's fine to have NAT and allocate from private ranges /24 nets for node/pods. But we shouldn't be trying to fit those practices to IPv6.

In IPv6 world, it is normal practice to have /64 on each link (interface), and in my opinion it is something that we need to target for every node allocation. Similarly, most of providers who implemented IPv6 for subscribers has rules: /64 is minimum allocation, /56 usually delivered via IPv6-PD mechanisms to end-user subscribers (256 of /64 nets) and /48 allocated for enterprise customers (65k of /64 nets). Even if provider does not provide IPv6 range, it is trivial to get /48 block via tunnel brokers like HE.net. Even with old but still functioning 6to4 technology, it is possible for every single IPv4 address in the world to get /48 block of IPv6 addresses that people can use.

So, to summarize my comments:
- In documentation and examples, don't use subnets which are smaller than /64 (/65-/128).
- In tests, use subnets from range `2001:db8::/32` which is explicitly allocated in rfc3849 to be used for documentation and code examples purposes.
- for scenarios where user for some reason wants to have networks for pods which are smaller than /64 (/65-/128),  let's round splitting node/pods part to border of 8 bit, for readability of addresses by humans.

I'd suggest we also get opinions from somebody from SIG-Network for that.
cc @kubernetes/sig-network-pr-reviews

**Comment by pmichali** [comment]
Thanks for the comments @kad. I agree with your points as goals for IPv6. Four points I'm trying to make here are:

1) Currently, it is broken and there is no valid pod network CIDR that can be configured for IPv6. Because of the restriction that the pod network CIDR be /66 or larger and a fixed node CIDR of /64, there is no valid configuration allowed. Until this restriction is eliminated, or we allow the node CIDR to be larger that /66, we cannot do IPv6 when a pod configuration is specified. This PR allows larger node CIDR to allow valid combinations to be specified from smaller pod networks.

2) Having a fixed value for node CIDR (/64) is inflexible, and prevents support for pod networks that are smaller than a /63. A perfectly reasonable pod subnet of /80 cannot be specified. This PR is calculating the CIDR, instead of hard coding it.

3) There is a limitation with the bit map used in Kubernetes in this area, which limits the max number of nodes to 32K (16 bits). Until a compressed bit map can be implemented (likely will be a long time), we must make sure that a calculated node CIDR does not violate that restriction. The poses some questions on how to best do that (maximize the number of pods/node, use a blended approach like the PR, etc).

4) The pod network CIDR is a configured setting. One can choose human readable value (and get a node CIDR that is also readable), or not, but we don't have control over what they configure. The question becomes, whether or not we should force human readable node CIDRs in the calculation (pick multiples of 8 that don't violate item [3]).

My hope here is that we can put in something that provides a reasonable solution so that we can move forward in the short term.

**Review by leblancd — COMMENTED** [review]
This change looks okay to me for a short term fix. In retrospect, I would have preferred other issues getting addressed first if practicable, e.g.:
* Allow for pod subnet sizes (and therefore node CIDRs) of /64 or larger (64 bit prefix or shorter); that is, make it less restrictive than the current limit on /66 prefixes for pod subnet. Part of removing this restriction would be to use the /math/big library to allow for potential 128-bit size values returned by RangeSize() here: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/ipallocator/allocator.go#L263. I don't know if there's also a bitmap of individual IP allocations that would need to be compressed (the other comments on bitmap compression are for allocating node CIDRs, not individual IPs).
* Allowing the calculation/configuration of node CIDRs to be optionally disabled if we're using CNI IPAM. Seems to me that most CNI IPAM plugins should manage how the pod subnet space is carved up between nodes, without kubernetes proper intervention, unless I'm missing something?
But even if the above 2 bullet points are feasible and were implemented, we'd still need the changes proposed in this PR, IMO (only the comments about the /66 restriction would need to be modified).

**Comment by rpothier** [comment]
This change looks good to me, I agree the root problem of the /66 requirement should be fixed as well.

**Comment by kad** [comment]
@pmichali @rpothier can you point me please, where do we have restriction for /66 ?

**Comment by pmichali** [comment]
Folks, I'll change the existing network in the test to use the documentation network.

I'll modify the algorithm so that the node CIDR to always be a multiple of 8, rather than trying to maximize the size of pods/node and number of nodes.

**Comment by pmichali** [comment]
@kad @rpothier @leblancd @yujuhong @timothysc Please check out the new change set.

This has the new algorithm, and I changed the comments to indicate there is a restriction of /66+ for pod network, but the code here will allow using smaller values. That way, when the restriction is removed, the logic will still work. Noted this in one of the test cases too, so that people know there is a limitation currently.

**Comment by leblancd** [comment]
@kad - The restriction that IPv6 pod subnets be /66 or smaller space (larger prefix) is set in place here: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/core/service/ipallocator/allocator.go#L265
```
// RangeSize returns the size of a range in valid addresses.
func RangeSize(subnet *net.IPNet) int64 {
	ones, bits := subnet.Mask.Size()
	if bits == 32 && (bits-ones) >= 31 || bits == 128 && (bits-ones) >= 63 {
		return 0
	}
	max := int64(1) << uint(bits-ones)
	return max
}
```
Since 'max' is a 64-bit signed int, the largest positive, power-of-2 value that it can have is 2^^62 (same as the size of a /66 subnet).

**Comment by kad** [comment]
@leblancd Thanks. didn't noticed that.
@pmichali in general looks good. Just one suggestion: let's not support something smaller than /112 - this will give 1 byte for addressing nodes and 1 byte for addressing pods as minimum. Then, rest of space split such that it is aligned on 8 bit border, but equally split between pods/nodes. With current formula we get situation where we can have huge amounts of pods, but minimum nodes (e.g. /66 pod CIDR  will result to /72 CIDR for node, means 0x3f nodes and 0xffffffffffffff pods, if my calculations are correct)

**Comment by pmichali** [comment]
@kad I can modify the algorithm to allocate more to the nodes, however, it still will be limited to 16 bits at most, so /66 would be /80, giving 48 bits for pods, and 14 bits for nodes.

**Comment by kad** [comment]
@pmichali sounds good. Just please make sure that across all possible values (/66 to /112) we have fair amount for nodes and pods. E.g. it should not be case like 16 bit for pods and only 2-3 bits for nodes.

**Comment by pmichali** [comment]
@kad Will push up a new version in a few mins. It will try to allow between 512-64k nodes (skewing to the upper side of the available bits, 9-16, instead of the lower side, 2-9).

File: cmd/kubeadm/app/phases/controlplane/manifests.go (hunk 1 above)
Comment (yujuhong) [reply cmd/kubeadm/app/phases/controlplane/manifests.go]: Changing maximum pods per node is a capacity has other implications, e.g., all resources on the node are sufficient to host 100 pods. I don't think in this case we need to change the default in this case. Users should/can configure their kubelets if they want a higher limit to match the cidr.
